mm2src/coins/utxo/utxo_common.rs

```diff
@@ -58,126 +61,266 @@ lazy_static! {
-        UtxoArcBuilder {
-            ctx,
-            ticker,
-            conf,
-            activation_params,
-            priv_key,
-        }
+pub async fn get_tx_fee(coin: &UtxoCoinFields) -> Result<ActualTxFee, JsonRpcError> {
+    let conf = &coin.conf;
+    match &coin.tx_fee {
+        TxFee::Dynamic(method) => {
+            let fee = coin
+                .rpc_client
+                .estimate_fee_sat(coin.decimals, method, &conf.estimate_fee_mode, conf.estimate_fee_blocks)
+                .compat()
+                .await?;
+            Ok(ActualTxFee::Dynamic(fee))
+        },
+        TxFee::FixedPerKb(satoshis) => Ok(ActualTxFee::FixedPerKb(*satoshis)),
     }
+}
 
-    pub fn with_priv_key(
-        ctx: &'a MmArc,
-        ticker: &'a str,
-        conf: &'a Json,
-        activation_params: UtxoActivationParams,
-        priv_key: &'a [u8],
-    ) -> UtxoArcBuilder<'a> {
-        UtxoArcBuilder {
-            ctx,
-            ticker,
-            conf,
-            activation_params,
-            priv_key: PrivKeyBuildPolicy::PrivKey(priv_key),
-        }
-    }
+pub fn derive_address<T>(
+    coin: &T,
+    hd_account: &UtxoHDAccount,
+    address_id: u32,
+    change: bool,
+) -> MmResult<HDAddress<Address>, Bip32Error>
+where
+    T: UtxoCommonOps,
+{
+    let change_child = ChildNumber::from(change as u32);
+    let address_id_child = ChildNumber::from(address_id);
+
+    let derived_pubkey = hd_account
+        .extended_pubkey
+        .derive_child(change_child)?
+        .derive_child(address_id_child)?;
+    let address = coin.address_from_extended_pubkey(&derived_pubkey);
+
+    let mut derivation_path = hd_account.account_derivation_path.clone();
+    derivation_path.push(change_child);
+    derivation_path.push(address_id_child);
+    Ok(HDAddress {
+        address,
+        derivation_path,
+    })
 }
 
-#[async_trait]
-impl UtxoCoinBuilder for UtxoArcBuilder<'_> {
-    type ResultCoin = UtxoArc;
+pub async fn hd_wallet_balance<T>(coin: &T, hd_wallet: &UtxoHDWallet) -> BalanceResult<Vec<HDAccountBalance>>
+where
+    T: UtxoCommonOps
+        + UtxoHDWalletOps
+        + HDWalletBalanceOps<HDWallet = UtxoHDWallet, HDAccount = UtxoHDAccount, Address = Address>,
+{
+    let orig_accounts = hd_wallet.accounts.lock().clone();
+    let mut accounts = orig_accounts.clone();
+    let mut balances = Vec::with_capacity(accounts.len());
 
-    async fn build(self) -> UtxoCoinBuildResult<Self::ResultCoin> {
-        let utxo = self.build_utxo_fields().await?;
-        Ok(UtxoArc(Arc::new(utxo)))
+    for hd_account in accounts.iter_mut() {
+        balances.push(coin.hd_account_balance(hd_wallet, hd_account).await?);
     }
 
-    fn ctx(&self) -> &MmArc { self.ctx }
+    // Update the accounts container if it hasn't been changed already while we requested the balances.
+    let mut current_accounts = hd_wallet.accounts.lock();
+    if *current_accounts == orig_accounts {
+        *current_accounts = accounts;
```

Comment:
Mutating the state looks like an unexpected side-effect to me in the fn that should be read-only by its name. Can we avoid it here?

Reply:
If we check addresses on the coin initialization only, it makes sense!

Reply:
Now there is `HDWalletCoinOps::apply_account_changes` that changes the state of the coin.
But `check_hd_account_balance` mutates the state at the end of the execution. Do you think it's acceptable?